plugins/optimization-detective/helper.php

```diff
@@ -60,6 +60,108 @@ function od_generate_media_query( ?int $minimum_viewport_width, ?int $maximum_vi
+	// > Access to script at '.../detect.js?ver=0.4.1' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
+	// So it's better to just avoid attempting to optimize Post Embed responses (which don't need optimization anyway).
+	if ( is_embed() ) {
+		$disabled_flags['is_embed'] = true;
+	}
+
+	// Skip posts that aren't published yet.
+	if ( is_preview() ) {
+		$disabled_flags['is_preview'] = true;
+	}
+
+	// Since injection of inline-editing controls interfere with breadcrumbs, while also just not necessary in this context.
+	if ( is_customize_preview() ) {
+		$disabled_flags['is_customize_preview'] = true;
+	}
+
+	// Since the images detected in the response body of a POST request cannot, by definition, be cached.
+	if ( isset( $_SERVER['REQUEST_METHOD'] ) && 'GET' !== $_SERVER['REQUEST_METHOD'] ) {
+		$disabled_flags['not_get_request'] = true;
+	}
+
+	// Page caching plugins can only reliably be told to invalidate a cached page when a post is available to trigger
+	// the relevant actions on.
+	if ( null === od_get_cache_purge_post_id() ) {
+		$disabled_flags['no_cache_purge_post_id'] = true;
+	}
+
+	// Check if any flags are set to true.
+	$has_disabled_flags = count( array_filter( $disabled_flags ) ) > 0;
+
+	/**
+	 * Filters whether the current response can be optimized.
+	 *
+	 * @since 0.1.0
+	 * @since n.e.x.t Added $disabled_flags parameter
+	 * @link https://github.com/WordPress/performance/blob/trunk/plugins/optimization-detective/docs/hooks.md#:~:text=Filter%3A%20od_can_optimize_response
+	 *
+	 * @param bool                $can_optimize    Whether response can be optimized.
+	 * @param array<string, bool> $disabled_flags  Flags indicating which conditions are disabling optimization.
+	 */
+	$can_optimize = (bool) apply_filters( 'od_can_optimize_response', ! $has_disabled_flags, $disabled_flags );
+
+	$reasons = array();
+	if ( ! $can_optimize ) {
+		$reason_messages = array(
+			'is_search'              => __( 'Page is not optimized because it is a search results page.', 'optimization-detective' ),
+			'is_embed'               => __( 'Page is not optimized because it is an embed.', 'optimization-detective' ),
+			'is_preview'             => __( 'Page is not optimized because it is a preview.', 'optimization-detective' ),
+			'is_customize_preview'   => __( 'Page is not optimized because it is a customize preview.', 'optimization-detective' ),
+			'not_get_request'        => __( 'Page is not optimized because it is not a GET request.', 'optimization-detective' ),
+			'no_cache_purge_post_id' => __( 'Page is not optimized because there is no post ID available for cache purging.', 'optimization-detective' ),
+		);
+
+		foreach ( $disabled_flags as $flag => $is_disabled ) {
+			if ( $is_disabled && isset( $reason_messages[ $flag ] ) ) {
+				$reasons[ $flag ] = $reason_messages[ $flag ];
+			}
+		}
```

Comment:
Alternative which uses `wp_array_slice_assoc()`:

```suggestion
		$reasons = wp_array_slice_assoc( $reason_messages, array_keys( array_filter( $disabled_flags ) ) );
```